src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshallingTests.cs

```diff
@@ -0,0 +1,1072 @@
+
+        [Fact]
+        public async Task FunctionPointerWithUnmanagedValueTypeParameter_Does_Not_Emit_Diagnostic()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public static unsafe void Test(delegate* unmanaged<ValueType, void> cb)
+    {
+        cb(default);
+    }
+}
+
+struct ValueType
+{
+    char c;
+}
+";
+            await VerifyCSAnalyzerAsync(source, allowUnsafeBlocks: true);
+        }
+
+        [Fact]
+        public async Task FunctionPointerWithUnmanagedValueTypeParameter_WithAutoLayout_Emits_Diagnostic()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public static unsafe void Test(delegate* unmanaged<ValueType, void> cb)
+    {
+        {|CA1420:cb(default)|};
+    }
+}
+
+[StructLayout(LayoutKind.Auto)]
+struct ValueType
+{
+    char c;
+}
+";
+            await VerifyCSAnalyzerAsync(source, allowUnsafeBlocks: true);
+        }
+
+        [Fact]
+        public async Task FunctionPointerWithUnmanagedValueTypeParameter_WithAutoLayoutField_Emits_Diagnostic()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public static unsafe void Test(delegate* unmanaged<ValueType, void> cb)
+    {
+        {|CA1420:cb(default)|};
+    }
+}
+
+struct ValueType
+{
+    ValueType2 v;
+}
+
+[StructLayout(LayoutKind.Auto)]
+struct ValueType2
+{
+    char c;
+}
+";
+            await VerifyCSAnalyzerAsync(source, allowUnsafeBlocks: true);
+        }
+
+        [Fact]
+        public async Task ManagedFunctionPointerWithUnmanagedValueTypeParameter_WithAutoLayoutField_Does_Not_Emit_Diagnostic()
+        {
+            string source = @"
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public static unsafe void Test(delegate*<ValueType, void> cb)
+    {
+        cb(default);
+    }
+}
+
+struct ValueType
+{
+    ValueType2 v;
+}
+
+[StructLayout(LayoutKind.Auto)]
+struct ValueType2
+{
+    char c;
+}
+";
+            await VerifyCSAnalyzerAsync(source, allowUnsafeBlocks: true);
+        }
+
+        [Fact]
+        public async Task MarshalOffsetOf_Emits_Diagnostic()
+        {
+            string source = @"
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+[assembly:DisableRuntimeMarshalling]
+
+class Foo
+{
+    public void Test()
+    {
+        IntPtr offset = {|CA1421:Marshal.OffsetOf(typeof(ValueType), ""field"")|};
+        IntPtr offsetGeneric = {|CA1421:Marshal.OffsetOf<ValueType>(""field"")|};
+    }
+}
+
+struct ValueType
+{
+    int field;
+}
+";
+            await VerifyCSAnalyzerAsync(source);
```

Comment:
Given CA1421 has a code fix, we should `VerifyCSCodeFixAsync` here, even if the scenario has no applicable code fix and we are forced to use `VerifyCSCodeFixAsync(source, source);`

Reply:
The definition of `VerifyCSAnalyzerAsync` in this file calls `VerifyCSCodeFixAsync(source, source)` on line 1047.